diff --git a/configure.ac b/configure.ac
index 3dadf542c..071f4e0f3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -4169,7 +4169,7 @@ case "$host" in
         VERSION_SOURCES="$srcdir/src/main/windows/*.rc"
         SDLMAIN_SOURCES="$srcdir/src/main/windows/*.c"
         SDL_CFLAGS="$SDL_CFLAGS -Dmain=SDL_main"
-        SDL_LIBS="-lSDL2main $SDL_LIBS -mwindows"
+        SDL_LIBS="-lSDL2main $SDL_LIBS -mconsole"
 
         # Check to see if this is a mingw or cygwin build
         have_mingw32=
