From 53d5dd1cd894e87bbd9f0162d96f05803773cafc Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sat, 6 Feb 2021 23:22:32 +0100
Subject: [PATCH 2/2] Add arm64 support to osxcross-macports

---
 tools/osxcross-macports | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tools/osxcross-macports b/tools/osxcross-macports
index e3f88b0..208121b 100755
--- a/tools/osxcross-macports
+++ b/tools/osxcross-macports
@@ -586,6 +586,8 @@ main()
         ARCH="i386"
       elif [ $opt == "-universal" -o $opt == "--universal" ]; then
         ARCH="i386-x86_64"
+      elif [ $opt == "--arm64" ]; then
+        ARCH="arm64"
       elif [ $opt == "-sm" -o $opt == "--select-mirror" ]; then
         selectMirror
         exit
-- 
2.25.1

