From 4b33d2be104bd38ab0e9b6c80bab0d5dd1f63777 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Wed, 1 Jun 2022 20:58:40 +0200
Subject: [PATCH 3/3] Add support for LLVM 14.x

---
 build_compiler_rt.sh | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/build_compiler_rt.sh b/build_compiler_rt.sh
index 17d91af..8630cbc 100755
--- a/build_compiler_rt.sh
+++ b/build_compiler_rt.sh
@@ -53,8 +53,9 @@ case $CLANG_VERSION in
   11.* ) BRANCH=release/11.x;    USE_CMAKE=1; ;;
   12.* ) BRANCH=release/12.x;    USE_CMAKE=1; ;;
   13.* ) BRANCH=release/13.x;    USE_CMAKE=1; ;;
-  14.* ) BRANCH=main;            USE_CMAKE=1; ;;
-     * ) echo "Unsupported Clang version, must be >= 3.2 and <= 14.0" 1>&2; exit 1;
+  14.* ) BRANCH=release/14.x;    USE_CMAKE=1; ;;
+  15.* ) BRANCH=main;            USE_CMAKE=1; ;;
+     * ) echo "Unsupported Clang version, must be >= 3.2 and <= 15.0" 1>&2; exit 1;
 esac
 
 if [ $(osxcross-cmp $CLANG_VERSION ">=" 3.5) -eq 1 ]; then
-- 
2.32.0

