From c933899097b323a55e264627099ee29e5378dfae Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Fri, 19 Mar 2021 12:18:07 +0100
Subject: [PATCH 2/4] Make target directory and triple configurable

---
 usage_examples/ios_toolchain/build.sh | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/usage_examples/ios_toolchain/build.sh b/usage_examples/ios_toolchain/build.sh
index e175323..860448f 100755
--- a/usage_examples/ios_toolchain/build.sh
+++ b/usage_examples/ios_toolchain/build.sh
@@ -87,8 +87,12 @@ if [ $# -lt 2 ]; then
     exit 1
 fi
 
+if [ -z "$TRIPLE" ]; then
 TRIPLE="arm-apple-darwin11"
+fi
+if [ -z "$TARGETDIR" ]; then
 TARGETDIR="$PWD/target"
+fi
 SDKDIR="$TARGETDIR/SDK"
 
 if [ -d $TARGETDIR ]; then
-- 
2.25.1

