Index: cctools-port/usage_examples/ios_toolchain/build.sh
===================================================================
--- cctools-port.orig/usage_examples/ios_toolchain/build.sh
+++ cctools-port/usage_examples/ios_toolchain/build.sh
@@ -169,7 +169,7 @@ rm -rf tmp
 
 mkdir -p tmp
 pushd tmp &>/dev/null
-git_clone_repository https://github.com/tpoechtrager/ldid.git master
+git_clone_repository https://github.com/tpoechtrager/ldid.git $LDID_VERSION
 pushd ldid &>/dev/null
 make INSTALLPREFIX=$TARGETDIR -j$JOBS install
 popd &>/dev/null
