Index: cctools-port/usage_examples/ios_toolchain/build.sh
===================================================================
--- cctools-port.orig/usage_examples/ios_toolchain/build.sh
+++ cctools-port/usage_examples/ios_toolchain/build.sh
@@ -88,7 +88,9 @@ if [ $# -lt 2 ]; then
 fi
 
 TRIPLE="arm-apple-darwin11"
+if [ -z "$TARGETDIR" ]; then
 TARGETDIR="$PWD/target"
+fi
 SDKDIR="$TARGETDIR/SDK"
 
 if [ -d $TARGETDIR ]; then
