Index: cctools-port/usage_examples/ios_toolchain/build.sh
===================================================================
--- cctools-port.orig/usage_examples/ios_toolchain/build.sh
+++ cctools-port/usage_examples/ios_toolchain/build.sh
@@ -153,10 +153,14 @@ elif ! which dsymutil &>/dev/null; then
     echo "int main(){return 0;}" | cc -xc -O2 -o $TARGETDIR/bin/dsymutil -
 fi
 
+if [ -z "${IPHONEOS_DEPLOYMENT_TARGET}" ]; then
+	IPHONEOS_DEPLOYMENT_TARGET=$SDK_VERSION
+fi
+
 verbose_cmd cc -O2 -Wall -Wextra -pedantic wrapper.c \
     -DSDK_DIR=\"\\\"$WRAPPER_SDKDIR\\\"\" \
     -DTARGET_CPU=\"\\\"$2\\\"\" \
-    -DOS_VER_MIN=\"\\\"$SDK_VERSION\\\"\" \
+    -DOS_VER_MIN=\"\\\"$IPHONEOS_DEPLOYMENT_TARGET\\\"\" \
     -o $TARGETDIR/bin/$TRIPLE-clang
 
 pushd $TARGETDIR/bin &>/dev/null
