diff --git a/scripts/026-libmpeg2.sh b/scripts/026-libmpeg2.sh
index f9bcebe..2300b8e 100755
--- a/scripts/026-libmpeg2.sh
+++ b/scripts/026-libmpeg2.sh
@@ -3,7 +3,7 @@
 VER=0.5.1
 
 ## Download the source code.
-if [ ! -f libmpeg2-${VER}.tar.gz ]; then wget --continue https://libmpeg2.sourceforge.io/files/libmpeg2-${VER}.tar.gz; fi
+if [ ! -f libmpeg2-${VER}.tar.gz ]; then wget --continue https://web.archive.org/web/https://libmpeg2.sourceforge.io/files/libmpeg2-${VER}.tar.gz; fi
 
 ## Download an up-to-date config.guess and config.sub
 if [ ! -f config.guess ]; then wget --continue http://git.savannah.gnu.org/cgit/config.git/plain/config.guess; fi
