diff --git a/config/psptoolchain-allegrex-config.sh b/config/psptoolchain-allegrex-config.sh
index e71842a..29d5c80 100644
--- a/config/psptoolchain-allegrex-config.sh
+++ b/config/psptoolchain-allegrex-config.sh
@@ -1,13 +1,14 @@
 #!/bin/bash
 
+# Don't use the override file as conflicts in this file will allow us to track branch changes
 PSPTOOLCHAIN_ALLEGREX_BINUTILS_REPO_URL="https://github.com/pspdev/binutils-gdb.git"
-PSPTOOLCHAIN_ALLEGREX_BINUTILS_DEFAULT_REPO_REF="allegrex-v2.44"
+PSPTOOLCHAIN_ALLEGREX_BINUTILS_DEFAULT_REPO_REF="$BINUTILS_VERSION"
 PSPTOOLCHAIN_ALLEGREX_GCC_REPO_URL="https://github.com/pspdev/gcc.git"
-PSPTOOLCHAIN_ALLEGREX_GCC_DEFAULT_REPO_REF="allegrex-v15.1.1"
+PSPTOOLCHAIN_ALLEGREX_GCC_DEFAULT_REPO_REF="$GCC_VERSION"
 PSPTOOLCHAIN_ALLEGREX_NEWLIB_REPO_URL="https://github.com/pspdev/newlib.git"
-PSPTOOLCHAIN_ALLEGREX_NEWLIB_DEFAULT_REPO_REF="allegrex-v4.5.0"
+PSPTOOLCHAIN_ALLEGREX_NEWLIB_DEFAULT_REPO_REF="$NEWLIB_VERSION"
 PSPTOOLCHAIN_ALLEGREX_PTHREAD_EMBEDDED_REPO_URL="https://github.com/pspdev/pthread-embedded.git"
-PSPTOOLCHAIN_ALLEGREX_PTHREAD_EMBEDDED_DEFAULT_REPO_REF="platform_agnostic"
+PSPTOOLCHAIN_ALLEGREX_PTHREAD_EMBEDDED_DEFAULT_REPO_REF="$PTHREAD_EMBEDDED_VERSION"
 
 if test -f "$PSPDEV_CONFIG_OVERRIDE"; then
   source "$PSPDEV_CONFIG_OVERRIDE"
diff --git a/scripts/001-binutils.sh b/scripts/001-binutils.sh
index af20a3d..58bb8ce 100755
--- a/scripts/001-binutils.sh
+++ b/scripts/001-binutils.sh
@@ -22,13 +22,11 @@ if test -n "$1"; then
   printf 'Using specified repo reference %s\n' "$REPO_REF"
 fi
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone --depth 1 -b "$REPO_REF" "$REPO_URL" "$REPO_FOLDER"
-else
-  git -C "$REPO_FOLDER" fetch origin
-  git -C "$REPO_FOLDER" reset --hard "origin/$REPO_REF"
-  git -C "$REPO_FOLDER" checkout "$REPO_REF"
-fi
+mkdir "$REPO_FOLDER"
+git -C "$REPO_FOLDER" init
+git -C "$REPO_FOLDER" remote add origin "$REPO_URL"
+git -C "$REPO_FOLDER" fetch --depth 1 origin "$REPO_REF"
+git -C "$REPO_FOLDER" checkout FETCH_HEAD
 
 cd "$REPO_FOLDER"
 
diff --git a/scripts/002-gcc-stage1.sh b/scripts/002-gcc-stage1.sh
index d8f537d..adf0759 100755
--- a/scripts/002-gcc-stage1.sh
+++ b/scripts/002-gcc-stage1.sh
@@ -22,13 +22,11 @@ if test -n "$1"; then
   printf 'Using specified repo reference %s\n' "$REPO_REF"
 fi
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone --depth 1 -b "$REPO_REF" "$REPO_URL" "$REPO_FOLDER"
-else
-  git -C "$REPO_FOLDER" fetch origin
-  git -C "$REPO_FOLDER" reset --hard "origin/$REPO_REF"
-  git -C "$REPO_FOLDER" checkout "$REPO_REF"
-fi
+mkdir "$REPO_FOLDER"
+git -C "$REPO_FOLDER" init
+git -C "$REPO_FOLDER" remote add origin "$REPO_URL"
+git -C "$REPO_FOLDER" fetch --depth 1 origin "$REPO_REF"
+git -C "$REPO_FOLDER" checkout FETCH_HEAD
 
 cd "$REPO_FOLDER"
 
diff --git a/scripts/003-newlib.sh b/scripts/003-newlib.sh
index 25e8e43..51bb7af 100755
--- a/scripts/003-newlib.sh
+++ b/scripts/003-newlib.sh
@@ -22,13 +22,11 @@ if test -n "$1"; then
   printf 'Using specified repo reference %s\n' "$REPO_REF"
 fi
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone --depth 1 -b "$REPO_REF" "$REPO_URL" "$REPO_FOLDER"
-else
-  git -C "$REPO_FOLDER" fetch origin
-  git -C "$REPO_FOLDER" reset --hard "origin/$REPO_REF"
-  git -C "$REPO_FOLDER" checkout "$REPO_REF"
-fi
+mkdir "$REPO_FOLDER"
+git -C "$REPO_FOLDER" init
+git -C "$REPO_FOLDER" remote add origin "$REPO_URL"
+git -C "$REPO_FOLDER" fetch --depth 1 origin "$REPO_REF"
+git -C "$REPO_FOLDER" checkout FETCH_HEAD
 
 cd "$REPO_FOLDER"
 
diff --git a/scripts/004-pthread-embedded.sh b/scripts/004-pthread-embedded.sh
index 04a70fa..da197e8 100755
--- a/scripts/004-pthread-embedded.sh
+++ b/scripts/004-pthread-embedded.sh
@@ -22,13 +22,11 @@ if test -n "$1"; then
   printf 'Using specified repo reference %s\n' "$REPO_REF"
 fi
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone --depth 1 -b "$REPO_REF" "$REPO_URL" "$REPO_FOLDER"
-else
-  git -C "$REPO_FOLDER" fetch origin
-  git -C "$REPO_FOLDER" reset --hard "origin/$REPO_REF"
-  git -C "$REPO_FOLDER" checkout "$REPO_REF"
-fi
+mkdir "$REPO_FOLDER"
+git -C "$REPO_FOLDER" init
+git -C "$REPO_FOLDER" remote add origin "$REPO_URL"
+git -C "$REPO_FOLDER" fetch --depth 1 origin "$REPO_REF"
+git -C "$REPO_FOLDER" checkout FETCH_HEAD
 
 cd "$REPO_FOLDER"
 
diff --git a/scripts/005-gcc-stage2.sh b/scripts/005-gcc-stage2.sh
index 9246fca..301717a 100755
--- a/scripts/005-gcc-stage2.sh
+++ b/scripts/005-gcc-stage2.sh
@@ -22,14 +22,6 @@ if test -n "$1"; then
   printf 'Using specified repo reference %s\n' "$REPO_REF"
 fi
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone --depth 1 -b "$REPO_REF" "$REPO_URL" "$REPO_FOLDER"
-else
-  git -C "$REPO_FOLDER" fetch origin
-  git -C "$REPO_FOLDER" reset --hard "origin/$REPO_REF"
-  git -C "$REPO_FOLDER" checkout "$REPO_REF"
-fi
-
 cd "$REPO_FOLDER"
 
 TARGET="psp"
