From 130429671a25d35be0fc88dddae11d9321d1b168 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sat, 30 Aug 2025 17:58:44 +0200
Subject: [PATCH 3/3] Generate all SDKs and not only MacOSX

---
 tools/gen_sdk_package.sh | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/tools/gen_sdk_package.sh b/tools/gen_sdk_package.sh
index fc24005..a240379 100755
--- a/tools/gen_sdk_package.sh
+++ b/tools/gen_sdk_package.sh
@@ -126,8 +126,8 @@ if [ -z "$XCODE_TOOLS" ]; then
 
   pushd "$XCODEDIR" &>/dev/null
 
-  if [ -d "Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs" ]; then
-    pushd "Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs" &>/dev/null
+  if [ -d "Contents/Developer/Platforms" ]; then
+    pushd "Contents/Developer/Platforms" &>/dev/null
   else
     if [ -d "../Packages" ]; then
       pushd "../Packages" &>/dev/null
@@ -191,12 +191,12 @@ else
 fi
 
 SDKS=()
-for pat in MacOS[0-9]*.sdk MacOSX[0-9]*.sdk; do
+for platform in $SDK_PLATFORMS; do
   while IFS= read -r d; do
     if [[ "$d" != "MacOS.sdk" && "$d" != "MacOSX.sdk" && "$d" != *Patch* && -d "$d" ]]; then
       SDKS+=("$d")
     fi
-  done < <(compgen -G "$pat")
+  done < <(compgen -G "${platform}.platform/Developer/SDKs/*.*.sdk")
 done
 
 
@@ -217,7 +217,8 @@ LIBCXXDIR3="usr/include/c++/v1"
 # Manual directory
 MANDIR="Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/share/man"
 
-for SDK in "${SDKS[@]}"; do
+for SDK_DIR in "${SDKS[@]}"; do
+  SDK=$(basename "$SDK_DIR")
   echo -n "packaging $(echo "$SDK" | sed -E "s/(.sdk|.pkg)//g") SDK "
   echo "(this may take several minutes) ..."
 
@@ -227,7 +228,7 @@ for SDK in "${SDKS[@]}"; do
   fi
 
   TMP=$(mktemp -d /tmp/XXXXXXXXXXX)
-  cp -r $(rreadlink $SDK) $TMP/$SDK &>/dev/null || true
+  cp -r $(rreadlink $SDK_DIR) $TMP/$SDK &>/dev/null || true
 
   pushd "$XCODEDIR" &>/dev/null
 
@@ -262,4 +263,4 @@ popd &>/dev/null
 popd &>/dev/null
 
 echo ""
-ls -lh | grep MacOSX
+ls -lh *.*.sdk.*
-- 
2.49.1

