From 8ed1c630ea783dc99be83eb0e7edc4d1cb45688d Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sun, 24 Jan 2021 19:19:14 +0100
Subject: [PATCH 3/3] Generate all SDKs and not only MacOSX

---
 tools/gen_sdk_package.sh | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/tools/gen_sdk_package.sh b/tools/gen_sdk_package.sh
index f29faec..d554c9d 100755
--- a/tools/gen_sdk_package.sh
+++ b/tools/gen_sdk_package.sh
@@ -127,8 +127,8 @@ if [ -z "$XCODE_TOOLS" ]; then
 
   pushd "$XCODEDIR" &>/dev/null
 
-  if [ -d "Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs" ]; then
-    pushd "Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs" &>/dev/null
+  if [ -d "Contents/Developer/Platforms" ]; then
+    pushd "Contents/Developer/Platforms" &>/dev/null
   else
     if [ -d "../Packages" ]; then
       pushd "../Packages" &>/dev/null
@@ -191,7 +191,7 @@ else
 
 fi
 
-SDKS=$(ls | grep -E "^MacOSX15.*|^MacOSX14.*|^MacOSX13.*|^MacOSX12.*|^MacOSX11.*|^MacOSX10.*" | grep -v "Patch")
+SDKS=$(for platform in $SDK_PLATFORMS; do ls -1d ${platform}.platform/Developer/SDKs/*.*.sdk | grep -v 'Patch'; done)
 
 if [ -z "$SDKS" ]; then
   echo "No SDK found" 1>&2
@@ -210,7 +210,8 @@ LIBCXXDIR3="usr/include/c++/v1"
 # Manual directory
 MANDIR="Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/share/man"
 
-for SDK in $SDKS; do
+for SDK_DIR in $SDKS; do
+  SDK=$(basename "$SDK_DIR")
   echo -n "packaging $(echo "$SDK" | sed -E "s/(.sdk|.pkg)//g") SDK "
   echo "(this may take several minutes) ..."
 
@@ -220,7 +221,7 @@ for SDK in $SDKS; do
   fi
 
   TMP=$(mktemp -d /tmp/XXXXXXXXXXX)
-  cp -r $(rreadlink $SDK) $TMP/$SDK &>/dev/null || true
+  cp -r $(rreadlink $SDK_DIR) $TMP/$SDK &>/dev/null || true
 
   pushd "$XCODEDIR" &>/dev/null
 
@@ -255,4 +256,4 @@ popd &>/dev/null
 popd &>/dev/null
 
 echo ""
-ls -lh | grep MacOSX
+ls -lh *.*.sdk.*
-- 
2.39.2

