commit 9ba0363c93eb3c2c7178f5a6b7a28466239c3eed
Author: Le Philousophe <lephilousophe@users.noreply.github.com>
Date:   Sun Aug 6 22:51:42 2023 +0200

    Fix Android build

diff --git a/Configurations/15-android.conf b/Configurations/15-android.conf
index 41ad9223e0..237b5f8735 100644
--- a/Configurations/15-android.conf
+++ b/Configurations/15-android.conf
@@ -224,7 +224,7 @@ my %targets = (
         inherit_from     => [ "android" ],
         bn_ops           => add("RC4_CHAR"),
         asm_arch         => 'armv4',
-        perlasm_scheme   => "void",
+        perlasm_scheme   => "linux",
     },
     "android-arm64" => {
         inherit_from     => [ "android" ],
