From 8937e868ea0cb3a23932857c5f5a5c82eb559389 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sun, 24 Jan 2021 19:15:52 +0100
Subject: [PATCH 1/3] Set versions

---
 tools/tools.sh | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/tools/tools.sh b/tools/tools.sh
index d8155fe..008714d 100644
--- a/tools/tools.sh
+++ b/tools/tools.sh
@@ -485,7 +485,7 @@ function git_clone_repository
 
   if [ ! -d "$project_name" ]; then
     if [ -n "$branch" ]; then
-      git clone $git_extra_opts --branch "$branch" "$url" "$project_name"
+      git clone $git_extra_opts "$url" "$project_name"
     else
       git clone $git_extra_opts "$url" "$project_name"
     fi
@@ -730,7 +730,7 @@ function build_xar()
 {
   pushd $BUILD_DIR &>/dev/null
 
-  get_sources https://github.com/tpoechtrager/xar.git master
+  get_sources https://github.com/tpoechtrager/xar.git $XAR_VERSION
 
   if [ $f_res -eq 1 ]; then
     pushd $CURRENT_BUILD_PROJECT_NAME/xar &>/dev/null
@@ -768,7 +768,7 @@ function build_p7zip()
 
 function build_pbxz()
 {
-  get_sources https://github.com/tpoechtrager/pbzx.git master
+  get_sources https://github.com/tpoechtrager/pbzx.git $PBZX_VERSION
 
   if [ $f_res -eq 1 ]; then
     pushd $CURRENT_BUILD_PROJECT_NAME &>/dev/null
-- 
2.54.0

