From 8bc8d72d889e17ecaf50311a264e8a394c8f324b Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sat, 17 Sep 2022 11:36:50 +0200
Subject: [PATCH 2/2] Handle MacPorts variants

---
 tools/osxcross-macports | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/tools/osxcross-macports b/tools/osxcross-macports
index 3e87735..f388914 100755
--- a/tools/osxcross-macports
+++ b/tools/osxcross-macports
@@ -275,13 +275,20 @@ verifyFileIntegrity()
 
 getPkgUrl()
 {
-  local pkgname="$1"
+  local pkgname="${1%+*}"
+  local pkgvariant="${1##*+}"
   local pkgversion
   local pkgs
   local pkg
 
   set +e
 
+  if [ "$pkgvariant" = "$1" ]; then
+	  pkgvariant='[^+]*\.'
+  else
+	  pkgvariant="+${pkgvariant}\\."
+  fi
+
   local pkg_info_url="https://ports.macports.org"
   pkg_info_url+="/api/v1/ports/$pkgname/?format=json"
   pkgversion=$(getFileStdout "$pkg_info_url" | \
@@ -308,11 +315,11 @@ getPkgUrl()
   done
 
   local pkg=$(echo "$pkgs" | \
-              grep "$pkgname-$pkgversion" | grep $OSXVERSION | grep $ARCH | \
+              grep "${pkgname}-${pkgversion}_[0-9]\\+${pkgvariant}" | grep $OSXVERSION | grep $ARCH | \
               uniq | tail -n1)
   if [ -z "$pkg" ]; then
     pkg=$(echo "$pkgs" | \
-          grep "$pkgname-$pkgversion" | grep $OSXVERSION | grep "noarch" | \
+          grep "${pkgname}-${pkgversion}_[0-9]\\+${pkgvariant}" | grep $OSXVERSION | grep "noarch" | \
           uniq | tail -n1)
   fi
   if [ -z "$pkg" ]; then
-- 
2.39.2

