From 19f1ae86ebf1b34a16d77af8df5c70f0b22bf9a4 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Tue, 25 Jun 2024 05:56:44 +0000
Subject: [PATCH 3/4] Disable ORC which uses C++17 features on older SDKs

---
 build_compiler_rt.sh | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/build_compiler_rt.sh b/build_compiler_rt.sh
index da64909..308ff7b 100755
--- a/build_compiler_rt.sh
+++ b/build_compiler_rt.sh
@@ -181,6 +181,12 @@ if [ $f_res -eq 1 ]; then
         echo ""
       fi
 
+      # Disable ORC runtime on older SDK as it requires C++17
+      if [ $(osxcross-cmp $SDK_VERSION "<=" 10.13) -eq 1 ] && \
+	 [ $(osxcross-cmp $CLANG_VERSION ">=" 16.0) -eq 1 ]; then
+         extra_cmake_flags+="-DCOMPILER_RT_BUILD_ORC=OFF "
+      fi
+
       mkdir $build_dir
       pushd $build_dir &>/dev/null
 
-- 
2.49.0

