diff --git a/build_compiler_rt.sh b/build_compiler_rt.sh
index e0b6b91..7836709 100755
--- a/build_compiler_rt.sh
+++ b/build_compiler_rt.sh
@@ -56,8 +56,9 @@ case $CLANG_VERSION in
   14.* ) BRANCH=release/14.x;    USE_CMAKE=1; ;;
   15.* ) BRANCH=release/15.x;    USE_CMAKE=1; ;;
   16.* ) BRANCH=release/16.x;    USE_CMAKE=1; ;;
-  17.* ) BRANCH=main;            USE_CMAKE=1; ;;
-  18.* ) BRANCH=main;            USE_CMAKE=1; ;;
+  17.* ) BRANCH=release/17.x;    USE_CMAKE=1; ;;
+  18.* ) BRANCH=release/18.x;    USE_CMAKE=1; ;;
+  19.* ) BRANCH=main;            USE_CMAKE=1; ;;
      * ) echo "Unsupported Clang version, must be >= 3.2 and <= 18.0" 1>&2; exit 1;
 esac
 
