diff --git a/build_compiler_rt.sh b/build_compiler_rt.sh
index 3c6df7a..28721f7 100755
--- a/build_compiler_rt.sh
+++ b/build_compiler_rt.sh
@@ -57,7 +57,7 @@ case $CLANG_VERSION in
   15.* ) BRANCH=release/15.x;    USE_CMAKE=1; ;;
   16.* ) BRANCH=release/16.x;    USE_CMAKE=1; ;;
   17.* ) BRANCH=release/17.x;    USE_CMAKE=1; ;;
-  18.* ) BRANCH=main;            USE_CMAKE=1; ;; # Does not build with release/18.x
+  18.* ) BRANCH=release/18.x;    USE_CMAKE=1; ;;
   19.* ) BRANCH=main;            USE_CMAKE=1; ;;
      * ) echo "Unsupported Clang version, must be >= 3.2 and <= 19.0" 1>&2; exit 1;
 esac
