From 88a7bf502f8024632b9b23e7f5285a82ed388427 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sun, 30 Oct 2022 17:26:18 +0100
Subject: [PATCH 1/2] Fix additional libs for SDK 54.16

---
 native-build/makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/native-build/makefile b/native-build/makefile
index 9d41410..e35900b 100644
--- a/native-build/makefile
+++ b/native-build/makefile
@@ -308,7 +308,7 @@ additionals-libs:
 	mkdir -p $(CROSS_PREFIX)/ppc-amigaos/SDK/local/common/include
 	mkdir -p /tmp/abcd
 	cd downloads && lha xw=/tmp/abcd SDK_$(SDK_VERSION).lha
-	cd /tmp/abcd/SDK_Install && lha x pthread.lha && lha x zlib-1.2.3.lha
+	cd /tmp/abcd/SDK_Install && lha x pthreads-53.12.lha
 	cp -Rf /tmp/abcd/SDK_Install/Local/* $(CROSS_PREFIX)/ppc-amigaos/SDK/local
 	rm -Rf /tmp/abcd
 
-- 
2.52.0

