From 76605b69039bfc627cf8c02dac19f1fc7a23f20f Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sun, 30 Oct 2022 17:26:18 +0100
Subject: [PATCH] Fix additional libs for SDK 54.16

---
 native-build/makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/native-build/makefile b/native-build/makefile
index e7e448e..e28a1c7 100644
--- a/native-build/makefile
+++ b/native-build/makefile
@@ -206,7 +206,7 @@ additionals-libs:
 	mkdir -p $(CROSS_PREFIX)/ppc-amigaos/SDK/local/common/include
 	mkdir -p /tmp/abcd
 	cd downloads && lha xw=/tmp/abcd SDK_$(SDK_VERSION).lha
-	cd /tmp/abcd/SDK_Install && lha x pthread.lha && lha x zlib-1.2.3.lha
+	cd /tmp/abcd/SDK_Install && lha x pthreads-53.12.lha
 	cp -Rf /tmp/abcd/SDK_Install/Local/* $(CROSS_PREFIX)/ppc-amigaos/SDK/local
 	rm -Rf /tmp/abcd
 
-- 
2.34.1

