diff --git a/native-build/makefile b/native-build/makefile
index 26d508a..2b540dd 100644
--- a/native-build/makefile
+++ b/native-build/makefile
@@ -36,7 +36,7 @@ endif
 
 GCC_BRANCH_NAME:=$(word 1, $(subst ., , $(GCC_VERSION)))
 
-SDK_URL=http://www.hyperion-entertainment.biz/index.php?option=com_registration&amp;view=download&amp;format=raw&amp;file=69&amp;Itemid=82
+SDK_URL=http://www.hyperion-entertainment.biz/index.php?option=com_registration&amp;view=download&amp;format=raw&amp;file=82&amp;Itemid=82
 SDK_VERSION=53.30
 
 # Native tools
