Index: glib-2.84.4/meson.build
===================================================================
--- glib-2.84.4.orig/meson.build
+++ glib-2.84.4/meson.build
@@ -982,7 +982,7 @@ endif
 osx_ldflags = []
 glib_have_carbon = false
 glib_have_cocoa = false
-if host_system == 'darwin'
+if false and host_system == 'darwin'
   add_languages('objc', native: false, required: true)
   objcc = meson.get_compiler('objc')
   if not objcc.compiles('''#include <AvailabilityMacros.h>
