From 1b4ec62e366f57a44dda0d67a0c1099945d0d26b Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sat, 16 Oct 2021 12:39:23 +0200
Subject: [PATCH 1/2] Remove need of ObjectiveC and deprecated Carbon

---
 meson.build | 11 ++---------
 1 file changed, 2 insertions(+), 9 deletions(-)

diff --git a/meson.build b/meson.build
index c8e986510..3246b909c 100644
--- a/meson.build
+++ b/meson.build
@@ -849,15 +849,8 @@ glib_have_os_x_9_or_later = false
 glib_have_carbon = false
 glib_have_cocoa = false
 if host_system == 'darwin'
-  add_languages('objc', native: false, required: true)
-  objcc = meson.get_compiler('objc')
-
-  add_project_arguments(objcc.get_supported_arguments(warning_objc_args), language: 'objc')
-
   # Mac OS X Carbon support
-  glib_have_carbon = objcc.compiles('''#include <Carbon/Carbon.h>
-                                       #include <CoreServices/CoreServices.h>''',
-                                    name : 'Mac OS X Carbon support')
+  glib_have_carbon = false
 
   if glib_have_carbon
     glib_conf.set('HAVE_CARBON', true)
@@ -869,7 +862,7 @@ if host_system == 'darwin'
   endif
 
   # Mac OS X Cocoa support
-  glib_have_cocoa = objcc.compiles('''#include <Cocoa/Cocoa.h>
+  glib_have_cocoa = cc.compiles('''#include <Cocoa/Cocoa.h>
                                       #ifdef GNUSTEP_BASE_VERSION
                                       #error "Detected GNUstep, not Cocoa"
                                       #endif''',
-- 
2.39.2

