From d4f6c101aa1f3cddb823b6bb697595626f84bd86 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sat, 16 Oct 2021 12:39:23 +0200
Subject: [PATCH 2/4] Remove need of ObjectiveC and deprecated Carbon

---
 meson.build | 11 +++--------
 1 file changed, 3 insertions(+), 8 deletions(-)

diff --git a/meson.build b/meson.build
index 7022e85ca..8d3bf08cc 100644
--- a/meson.build
+++ b/meson.build
@@ -730,15 +730,10 @@ glib_have_os_x_9_or_later = false
 glib_have_carbon = false
 glib_have_cocoa = false
 if host_system == 'darwin'
-  add_languages('objc')
-  objcc = meson.get_compiler('objc')
-
   osx_ldflags += ['-Wl,-framework,CoreFoundation']
 
   # Mac OS X Carbon support
-  glib_have_carbon = objcc.compiles('''#include <Carbon/Carbon.h>
-                                       #include <CoreServices/CoreServices.h>''',
-                                    name : 'Mac OS X Carbon support')
+  glib_have_carbon = false
 
   if glib_have_carbon
     glib_conf.set('HAVE_CARBON', true)
@@ -751,7 +746,7 @@ if host_system == 'darwin'
   endif
 
   # Mac OS X Cocoa support
-  glib_have_cocoa = objcc.compiles('''#include <Cocoa/Cocoa.h>
+  glib_have_cocoa = cc.compiles('''#include <Cocoa/Cocoa.h>
                                       #ifdef GNUSTEP_BASE_VERSION
                                       #error "Detected GNUstep, not Cocoa"
                                       #endif''',
@@ -765,7 +760,7 @@ if host_system == 'darwin'
   # FIXME: libgio mix C and objC source files and there is no way to reliably
   # know which language flags it's going to use to link. Add to both languages
   # for now. See https://github.com/mesonbuild/meson/issues/3585.
-  add_project_link_arguments(osx_ldflags, language : ['objc', 'c'])
+  add_project_link_arguments(osx_ldflags, language : ['c'])
 endif
 
 # Check for futex(2)
-- 
2.38.2

