From 9930def5534c13c812fbd13b53e5cfaa644feb86 Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sun, 17 Oct 2021 16:04:14 +0200
Subject: [PATCH 3/4] Fix intl build with iOS

---
 meson.build | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/meson.build b/meson.build
index 8d3bf08cc..85b14caf9 100644
--- a/meson.build
+++ b/meson.build
@@ -761,6 +761,9 @@ if host_system == 'darwin'
   # know which language flags it's going to use to link. Add to both languages
   # for now. See https://github.com/mesonbuild/meson/issues/3585.
   add_project_link_arguments(osx_ldflags, language : ['c'])
+elif host_system == 'ios'
+  osx_ldflags += ['-Wl,-framework,CoreFoundation']
+  add_project_link_arguments(osx_ldflags, language : ['c'])
 endif
 
 # Check for futex(2)
-- 
2.38.2

