diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9dca996..6cc709e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -72,6 +72,7 @@ mark_as_advanced ( LIB_SUFFIX )
 
 # Default install directory names
 include ( DefaultDirs )
+include ( GNUInstallDirs )
 
 # Basic C library checks
 include ( CheckSTDC )
diff --git a/include/CMakeLists.txt b/include/CMakeLists.txt
index 76a8f87..7973219 100644
--- a/include/CMakeLists.txt
+++ b/include/CMakeLists.txt
@@ -22,5 +22,5 @@
 add_subdirectory ( fluidsynth )
 
 IF (NOT MACOSX_FRAMEWORK )
-    install ( FILES fluidsynth.h DESTINATION ${INCLUDE_INSTALL_DIR} )
+    install ( FILES fluidsynth.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR} )
 ENDIF (NOT MACOSX_FRAMEWORK )
diff --git a/include/fluidsynth/CMakeLists.txt b/include/fluidsynth/CMakeLists.txt
index b146d13..d4c00b4 100644
--- a/include/fluidsynth/CMakeLists.txt
+++ b/include/fluidsynth/CMakeLists.txt
@@ -46,6 +46,6 @@ if ( NOT MACOSX_FRAMEWORK )
       ${include_HEADERS}
       ${CMAKE_CURRENT_BINARY_DIR}/version.h
     DESTINATION 
-      ${INCLUDE_INSTALL_DIR}/fluidsynth 
+      ${CMAKE_INSTALL_INCLUDEDIR}/fluidsynth
   )
 endif ( NOT MACOSX_FRAMEWORK )
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index c14622d..4e99dab 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -152,5 +152,5 @@ target_link_libraries ( libfluidsynth
 )
 
 install(TARGETS libfluidsynth
-  LIBRARY DESTINATION lib
-  ARCHIVE DESTINATION lib)
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
