diff --git a/AUTHORS b/AUTHORS
index 41f1253bda..66d9201c25 100644
--- a/AUTHORS
+++ b/AUTHORS
@@ -33,6 +33,8 @@ Authors, for copyright and licensing purposes.
  * Vadim Barkov
  * Willem van Schaik
  * Zhijie Liang
+ * Apple Inc.
+   - Zixu Wang
  * Arm Holdings
    - Richard Townsend
  * Google Inc.
diff --git a/pngpriv.h b/pngpriv.h
index 6c7280cf53..190eb85cbf 100644
--- a/pngpriv.h
+++ b/pngpriv.h
@@ -556,18 +556,8 @@
     */
 #  include <float.h>
 
-#  if (defined(__MWERKS__) && defined(macintosh)) || defined(applec) || \
-    defined(THINK_C) || defined(__SC__) || defined(TARGET_OS_MAC)
-   /* We need to check that <math.h> hasn't already been included earlier
-    * as it seems it doesn't agree with <fp.h>, yet we should really use
-    * <fp.h> if possible.
-    */
-#    if !defined(__MATH_H__) && !defined(__MATH_H) && !defined(__cmath__)
-#      include <fp.h>
-#    endif
-#  else
-#    include <math.h>
-#  endif
+#  include <math.h>
+
 #  if defined(_AMIGA) && defined(__SASC) && defined(_M68881)
    /* Amiga SAS/C: We must include builtin FPU functions when compiling using
     * MATH=68881
