From b38504eef6947e9aa113d41018f51e966ff972ca Mon Sep 17 00:00:00 2001
From: Paul Cercueil <paul@crapouillou.net>
Date: Sat, 10 May 2014 21:41:06 +0200
Subject: [PATCH 5/6] video: fbdev: Init EVDEV if enabled in config

---
 src/video/fbdev/SDL_fbvideo.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/video/fbdev/SDL_fbvideo.c b/src/video/fbdev/SDL_fbvideo.c
index 672815369..235621bf6 100644
--- a/src/video/fbdev/SDL_fbvideo.c
+++ b/src/video/fbdev/SDL_fbvideo.c
@@ -29,6 +29,10 @@
 #include "SDL_loadso.h"
 #include "SDL_events.h"
 
+#ifdef SDL_INPUT_LINUXEV
+#include "../../core/linux/SDL_evdev.h"
+#endif
+
 #include "SDL_fbvideo.h"
 #include "SDL_fbopengles.h"
 #include "../../events/SDL_mouse_c.h"
@@ -43,6 +47,9 @@ FB_Available(void)
 static void
 FB_PumpEvents(_THIS)
 {
+#ifdef SDL_INPUT_LINUXEV
+    SDL_EVDEV_Poll();
+#endif
 }
 
 static void
@@ -135,12 +142,20 @@ FB_VideoInit(_THIS)
     display.current_mode = current_mode;
 
     SDL_AddVideoDisplay(&display);
+
+#ifdef SDL_INPUT_LINUXEV
+    SDL_EVDEV_Init();
+#endif
+
     return 1;
 }
 
 void
 FB_VideoQuit(_THIS)
 {
+#ifdef SDL_INPUT_LINUXEV
+    SDL_EVDEV_Quit();
+#endif
 }
 
 void
-- 
2.25.1

