From b85c7449be81f0f707dca329c2fc67733d4c6856 Mon Sep 17 00:00:00 2001
From: Cameron Cawley <ccawley2011@gmail.com>
Date: Fri, 22 Dec 2017 20:04:35 +0000
Subject: [PATCH] Fix compilation with MinGW

---
 libfaad/decoder.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libfaad/decoder.c b/libfaad/decoder.c
index 52a387f..4f4b011 100644
--- a/libfaad/decoder.c
+++ b/libfaad/decoder.c
@@ -53,7 +53,7 @@
 uint16_t dbg_count;
 #endif
 
-#if defined WIN32 || defined _WIN32 || defined WIN64 || defined _WIN64
+#if (defined WIN32 || defined _WIN32 || defined WIN64 || defined _WIN64) && !defined(PACKAGE_VERSION)
 #include "win32_ver.h"
 #endif
 
