diff --git a/makerom/deps/libyaml/makefile b/makerom/deps/libyaml/makefile
index 785e3b8..a6dfba7 100644
--- a/makerom/deps/libyaml/makefile
+++ b/makerom/deps/libyaml/makefile
@@ -86,7 +86,7 @@ else ifeq ($(PROJECT_PLATFORM), GNU)
 	#CXX =
 	WARNFLAGS = -Wall -Wno-unused-value -Wno-unused-but-set-variable
 	ARCHFLAGS =
-	INC +=
+	INC += -D_GNU_SOURCE
 	LIB +=
 	ARFLAGS = cr -o
 else ifeq ($(PROJECT_PLATFORM), MACOS)
