diff --git a/scripts/001-allegrex.sh b/scripts/001-allegrex.sh
index 096b094..47ec86e 100755
--- a/scripts/001-allegrex.sh
+++ b/scripts/001-allegrex.sh
@@ -5,14 +5,15 @@
 REPO_URL="https://github.com/pspdev/psptoolchain-allegrex"
 REPO_FOLDER="psptoolchain-allegrex"
 
-# Checking if a specific TAG has been selected, it is passed using parameter $1
-[  -z "$1" ] && REPO_REFERENCE="main" || REPO_REFERENCE=$1
-echo "Using repo reference $REPO_REFERENCE"
+mkdir "$REPO_FOLDER" && cd "$REPO_FOLDER"
+git init || { exit 1; }
+git remote add origin "$REPO_URL" || { exit 1; }
+git fetch --depth 1 origin "$PSPTOOLCHAIN_ALLEGREX_VERSION" || { exit 1; }
+git checkout FETCH_HEAD || { exit 1; }
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone $REPO_URL -b "${REPO_REFERENCE}" || exit 1
-fi
-cd "$REPO_FOLDER" && git fetch origin && git reset --hard "origin/${REPO_REFERENCE}" && git checkout "${REPO_REFERENCE}" || exit 1
+for p in "$PACKAGE_DIR/$REPO_FOLDER-patches"/*; do
+	patch -p1 < "$p" || { exit 1; }
+done
 
 ## Build and install.
 ./toolchain.sh || { exit 1; }
diff --git a/scripts/002-extra.sh b/scripts/002-extra.sh
index ab89aee..5d5ec20 100755
--- a/scripts/002-extra.sh
+++ b/scripts/002-extra.sh
@@ -5,14 +5,15 @@
 REPO_URL="https://github.com/pspdev/psptoolchain-extra"
 REPO_FOLDER="psptoolchain-extra"
 
-# Checking if a specific TAG has been selected, it is passed using parameter $1
-[  -z "$1" ] && REPO_REFERENCE="main" || REPO_REFERENCE=$1
-echo "Using repo reference $REPO_REFERENCE"
+mkdir "$REPO_FOLDER" && cd "$REPO_FOLDER"
+git init || { exit 1; }
+git remote add origin "$REPO_URL" || { exit 1; }
+git fetch --depth 1 origin "$PSPTOOLCHAIN_EXTRA_VERSION" || { exit 1; }
+git checkout FETCH_HEAD || { exit 1; }
 
-if test ! -d "$REPO_FOLDER"; then
-  git clone $REPO_URL -b "${REPO_REFERENCE}" || exit 1
-fi
-cd "$REPO_FOLDER" && git fetch origin && git reset --hard "origin/${REPO_REFERENCE}" && git checkout "${REPO_REFERENCE}" || exit 1
+for p in "$PACKAGE_DIR/$REPO_FOLDER-patches"/*; do
+	patch -p1 < "$p" || { exit 1; }
+done
 
 ## Build and install.
-./build-all.sh || { exit 1; }
\ No newline at end of file
+./build-all.sh || { exit 1; }
