Index: Tools/oslib/os.h
===================================================================
--- a/Tools/oslib/os.h	(revision 455)
+++ b/Tools/oslib/os.h	(working copy)
@@ -29,6 +29,7 @@
 #endif
 
 #if defined EXECUTE_ON_UNIX
+  #include <stdio.h>
   #define OS_ERROR(e) (os_set_lasterr((e)))
 #endif
 
@@ -168,8 +169,12 @@
 typedef byte os_f;
       /*Deprecated 8-bit File handle. From OSLib V6.3 onwards the headers equate this name to its 32 bit equivalent, but the library retains this symbol for binary compatibility. To revert to legacy bahaviour, #define OSLIB_F8 during compilation.*/
 
+#ifdef EXECUTE_ON_UNIX
+typedef FILE* os_fw;
+#else
 typedef bits os_fw;
       /*Wide (32-bit) File handle*/
+#endif
 
 typedef byte os_gcol;
       /*Graphical colour, 6 or 8 bits*/
