Index: curl-7.64.0/configure
===================================================================
--- curl-7.64.0.orig/configure
+++ curl-7.64.0/configure
@@ -22121,7 +22121,7 @@ fi
 $as_echo_n "checking whether to enable Apple OS native SSL/TLS... " >&6; }
 if test -z "$ssl_backends" -o "x$OPT_DARWINSSL" != xno; then
   if test "x$OPT_DARWINSSL" != "xno" &&
-     test -d "/System/Library/Frameworks/Security.framework"; then
+     true; then
     { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
 $as_echo "yes" >&6; }
 
