From b808a8862786758fa4543f29020df4cd43a43c9d Mon Sep 17 00:00:00 2001
From: Le Philousophe <lephilousophe@users.noreply.github.com>
Date: Sat, 18 Oct 2025 17:02:02 +0200
Subject: [PATCH 3/3] Don't use pthread_setname_np when not available

---
 vpx_util/vpx_thread.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/vpx_util/vpx_thread.c b/vpx_util/vpx_thread.c
index 0d0e2f5..5c7d28e 100644
--- a/vpx_util/vpx_thread.c
+++ b/vpx_util/vpx_thread.c
@@ -18,6 +18,10 @@
 #define _GNU_SOURCE
 #endif
 
+#ifndef __GLIBC_PREREQ
+#define __GLIBC_PREREQ(M,m) 0
+#endif
+
 #include <assert.h>
 #include <string.h>  // for memset()
 #include "./vpx_config.h"
@@ -51,7 +55,7 @@ static THREADFN thread_loop(void *ptr) {
     thread_name[sizeof(thread_name) - 1] = '\0';
     pthread_setname_np(thread_name);
   }
-#elif (defined(__GLIBC__) && !defined(__GNU__)) || defined(__BIONIC__)
+#elif (defined(__GLIBC__) && !defined(__GNU__) && __GLIBC_PREREQ(2, 12)) || defined(__BIONIC__)
   if (worker->thread_name != NULL) {
     // Linux and Android require names (with nul) fit in 16 chars, otherwise
     // pthread_setname_np() returns ERANGE (34).
-- 
2.49.1

