diff --git a/scripts/001-psptoolchain.sh b/scripts/001-psptoolchain.sh
index 732e96a..babb6f8 100755
--- a/scripts/001-psptoolchain.sh
+++ b/scripts/001-psptoolchain.sh
@@ -4,12 +4,16 @@
 ## Download the source code.
 REPO_URL="https://github.com/pspdev/psptoolchain"
 REPO_FOLDER="psptoolchain"
-BRANCH_NAME="master"
-if test ! -d "$REPO_FOLDER"; then
-	git clone --depth 1 -b $BRANCH_NAME $REPO_URL && cd $REPO_FOLDER || { exit 1; }
-else
-	cd $REPO_FOLDER && git fetch origin && git reset --hard origin/${BRANCH_NAME} || { exit 1; }
-fi
+
+mkdir "$REPO_FOLDER" && cd "$REPO_FOLDER"
+git init
+git remote add origin "$REPO_URL"
+git fetch --depth 1 origin "$PSPTOOLCHAIN_VERSION"
+git checkout FETCH_HEAD
+
+for p in "$PACKAGE_DIR/$REPO_FOLDER-patches"/*; do
+	patch -p1 < "$p"
+done
 
 ## Build and install.
 ./toolchain.sh || { exit 1; }
diff --git a/scripts/002-pspsdk.sh b/scripts/002-pspsdk.sh
index 629f3ac..fe11891 100755
--- a/scripts/002-pspsdk.sh
+++ b/scripts/002-pspsdk.sh
@@ -4,12 +4,12 @@
 ## Download the source code.
 REPO_URL="https://github.com/pspdev/pspsdk"
 REPO_FOLDER="pspsdk"
-BRANCH_NAME="master"
-if test ! -d "$REPO_FOLDER"; then
-	git clone --depth 1 -b $BRANCH_NAME $REPO_URL && cd $REPO_FOLDER || { exit 1; }
-else
-	cd $REPO_FOLDER && git fetch origin && git reset --hard origin/${BRANCH_NAME} || { exit 1; }
-fi
+
+mkdir "$REPO_FOLDER" && cd "$REPO_FOLDER"
+git init
+git remote add origin "$REPO_URL"
+git fetch --depth 1 origin "$PSPSDK_VERSION"
+git checkout FETCH_HEAD
 
 ## Determine the maximum number of processes that Make can work with.
 PROC_NR=$(getconf _NPROCESSORS_ONLN)
diff --git a/scripts/004-psplinkusb-extra.sh b/scripts/004-psplinkusb-extra.sh
index df7bce4..2dc9058 100755
--- a/scripts/004-psplinkusb-extra.sh
+++ b/scripts/004-psplinkusb-extra.sh
@@ -4,12 +4,12 @@
 ## Download the source code.
 REPO_URL="https://github.com/pspdev/psplinkusb"
 REPO_FOLDER="psplinkusb"
-BRANCH_NAME="master"
-if test ! -d "$REPO_FOLDER"; then
-	git clone --depth 1 -b $BRANCH_NAME $REPO_URL && cd $REPO_FOLDER || { exit 1; }
-else
-	cd $REPO_FOLDER && git fetch origin && git reset --hard origin/${BRANCH_NAME} || { exit 1; }
-fi
+
+mkdir "$REPO_FOLDER" && cd "$REPO_FOLDER"
+git init
+git remote add origin "$REPO_URL"
+git fetch --depth 1 origin "$PSPLINKUSB_VERSION"
+git checkout FETCH_HEAD
 
 ## Determine the maximum number of processes that Make can work with.
 PROC_NR=$(getconf _NPROCESSORS_ONLN)
diff --git a/scripts/005-ebootsigner-extra.sh b/scripts/005-ebootsigner-extra.sh
index c6aff68..c79f456 100755
--- a/scripts/005-ebootsigner-extra.sh
+++ b/scripts/005-ebootsigner-extra.sh
@@ -4,12 +4,12 @@
 ## Download the source code.
 REPO_URL="https://github.com/pspdev/ebootsigner"
 REPO_FOLDER="ebootsigner"
-BRANCH_NAME="master"
-if test ! -d "$REPO_FOLDER"; then
-	git clone --depth 1 -b $BRANCH_NAME $REPO_URL && cd $REPO_FOLDER || { exit 1; }
-else
-	cd $REPO_FOLDER && git fetch origin && git reset --hard origin/${BRANCH_NAME} || { exit 1; }
-fi
+
+mkdir "$REPO_FOLDER" && cd "$REPO_FOLDER"
+git init
+git remote add origin "$REPO_URL"
+git fetch --depth 1 origin "$EBOOTSIGNER_VERSION"
+git checkout FETCH_HEAD
 
 ## Determine the maximum number of processes that Make can work with.
 PROC_NR=$(getconf _NPROCESSORS_ONLN)
