diff --git a/depends/check-dependencies.sh b/depends/check-dependencies.sh
index c97a4f7..cfa5181 100755
--- a/depends/check-dependencies.sh
+++ b/depends/check-dependencies.sh
@@ -67,7 +67,7 @@ function check_aclocal
 # check_header    gmp             gmp.h
 # check_header    mpc             mpc.h
 # check_header    mpfr            mpfr.h
-check_header    libelf          elf.h libelf.h libelf/libelf.h gelf.h libelf/gelf.h
+# check_header    libelf          elf.h libelf.h libelf/libelf.h gelf.h libelf/gelf.h
 check_header    libusb          usb.h
 check_header    ncurses         ncurses.h ncurses/ncurses.h
 check_header    zlib            zlib.h
@@ -86,7 +86,7 @@ check_program   make
 check_program   gcc
 check_program   g++
 
-check_program   sdl-config
+# check_program   sdl-config
 # check_program   freetype-config
 
 check_program   bison
