diff --git a/depends/check-dependencies.sh b/depends/check-dependencies.sh
index e873a74..a27c6a1 100755
--- a/depends/check-dependencies.sh
+++ b/depends/check-dependencies.sh
@@ -53,12 +53,12 @@ if [ "$(uname)" == "Darwin" ]; then
   header_paths+=("`xcrun --show-sdk-path`/usr/include")
 fi
 
-check_header    libelf          elf.h libelf.h libelf/libelf.h gelf.h libelf/gelf.h
+#check_header    libelf          elf.h libelf.h libelf/libelf.h gelf.h libelf/gelf.h
 check_header    libusb          usb.h
 check_header    ncurses         ncurses.h ncurses/ncurses.h
 check_header    zlib            zlib.h
-check_header    libcurl         curl/curl.h
-check_header    gpgme           gpgme.h
+#check_header    libcurl         curl/curl.h
+#check_header    gpgme           gpgme.h
 
 check_program   git
 check_program   svn
@@ -75,7 +75,7 @@ check_program   cmake
 check_program   make
 check_program   gcc
 check_program   g++
-check_program   m4
+#check_program   m4
 
 check_program   bison
 check_program   flex
@@ -86,7 +86,7 @@ check_program   which
 check_program   makeinfo
 check_program   doxygen
 
-check_program   python3
+#check_program   python3
 
 # Sometimes things will be a little different on Mac OS X...
 if [ "$(uname)" == "Darwin" ]; then
@@ -107,8 +107,8 @@ if [ "$(uname)" == "Darwin" ]; then
     # the one comes along with OS X.
     check_program      glibtoolize
 else
-    check_header       libarchive   archive.h
-    check_header       libssl       openssl/ssl.h
+#    check_header       libarchive   archive.h
+#    check_header       libssl       openssl/ssl.h
     check_header       readline     readline.h readline/readline.h
     check_program      libtoolize
 fi
