commit 04fc2f30d2a27c5f707830708cb345bdad4eeee1
Author: Tony Theodore <tonyt@logyst.com>
Date:   Sun Sep 8 18:22:33 2019 +1000

    Allow optional build of bins, docs, and tests

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 350a1dce..e88c1a9c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -105,6 +105,11 @@ option ( enable-native-dls "compile native DLS support (requires C++17)" on )
 
 set ( osal "glib" CACHE STRING "OS abstraction to use, provided by src/utils/fluid_sys_${osal}.*" )
 
+# Options for build components
+option ( build-bins "build binary programs" on )
+option ( build-docs "build docs" on )
+option ( build-tests "build tests" on )
+
 # Platform specific options
 if ( CMAKE_SYSTEM MATCHES "Linux" )
     option ( enable-systemd "compile systemd support (if it is available)" on )
@@ -887,8 +892,12 @@ endif ()
 
 # Process subdirectories
 add_subdirectory ( src )
-add_subdirectory ( test )
-add_subdirectory ( doc )
+if ( build-tests )
+    add_subdirectory ( test )
+endif ()
+if ( build-docs )
+    add_subdirectory ( doc )
+endif ()
 
 # pkg-config support
 set ( prefix "${CMAKE_INSTALL_PREFIX}" )
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 15017af9..fd9ba9e8 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -473,6 +473,8 @@ target_link_libraries ( libfluidsynth PRIVATE libfluidsynth-OBJ )
 
 # ************ CLI program ************
 
+if ( build-bins )
+    set ( fluidsynth_bin "fluidsynth" )
 set ( fluidsynth_SOURCES fluidsynth.c )
 
 if ( WASAPI_SUPPORT )
@@ -558,15 +560,17 @@ if ( MINGW AND enable-unicode )
     target_link_options( fluidsynth PRIVATE -municode )
 endif()
 
+endif ( build-bins )
+
 if ( MACOSX_FRAMEWORK )
-  install ( TARGETS fluidsynth libfluidsynth
+  install ( TARGETS ${fluidsynth_bin} libfluidsynth
     EXPORT FluidSynthTargets
     RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
     FRAMEWORK DESTINATION "${FRAMEWORK_INSTALL_PREFIX}/${FRAMEWORK_INSTALL_DIR}"
     ARCHIVE DESTINATION "${FRAMEWORK_INSTALL_PREFIX}/${FRAMEWORK_INSTALL_DIR}"
 )
 else ( MACOSX_FRAMEWORK )
-  install ( TARGETS fluidsynth libfluidsynth
+  install ( TARGETS ${fluidsynth_bin} libfluidsynth
     EXPORT FluidSynthTargets
     RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
     LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
