Index: flac-1.2.1/configure.in
===================================================================
--- flac-1.2.1.orig/configure.in
+++ flac-1.2.1/configure.in
@@ -226,9 +226,9 @@ AC_HELP_STRING([--disable-xmms-plugin],
 	*) AC_MSG_ERROR(bad value ${enableval} for --enable-xmms-plugin) ;;
 esac],[enable_xmms_plugin=true])
 if test "x$enable_xmms_plugin" != xfalse ; then
-	AM_PATH_XMMS(0.9.5.1, , AC_MSG_WARN([*** XMMS >= 0.9.5.1 not installed - XMMS support will not be built]))
+	AC_MSG_ERROR(No XMMS support)
 fi
-AM_CONDITIONAL(FLaC__HAS_XMMS, test -n "$XMMS_INPUT_PLUGIN_DIR")
+AM_CONDITIONAL(FLaC__HAS_XMMS, false)
 
 dnl build FLAC++ or not
 AC_ARG_ENABLE([cpplibs],
