commit 2c932128fea1b43342bcfcb1d7ea9da122ac7e23 (HEAD)
Author: Le Philousophe <lephilousophe@users.noreply.github.com>
Date:   Sat Sep 21 22:25:58 2019 +0200

    Don't use _FORTIFY_SOURCE when not needed

diff --git a/configure.ac b/configure.ac
index 070ab357..4971c583 100644
--- a/configure.ac
+++ b/configure.ac
@@ -400,7 +400,6 @@ if test x$ac_cv_c_compiler_gnu = xyes ; then
 	CXXFLAGS="$CXXFLAGS -Wall -Wextra -Wcast-align -Wshadow -Wwrite-strings -Wctor-dtor-privacy -Wnon-virtual-dtor -Wreorder -Wsign-promo -Wundef " # -Wcast-qual -Wbad-function-cast -Wwrite-strings -Woverloaded-virtual -Wmissing-declarations
 
 	XIPH_ADD_CFLAGS([-Wdeclaration-after-statement])
-	XIPH_ADD_CFLAGS([-D_FORTIFY_SOURCE=2])
 
 	AC_LANG_PUSH([C++])
 	XIPH_ADD_CXXFLAGS([-Weffc++])
