Index: getandpatch.sh
===================================================================
--- a/getandpatch.sh	(revision 376)
+++ b/getandpatch.sh	(working copy)
@@ -70,13 +70,13 @@
 	case $1 in
 
 	*glibc-200*gz)
-		wget --tries=5 -P ${TARBALLS_DIR} -c $1 || wget --tries=5 --passive-ftp -P ${TARBALLS_DIR} -c $1 || \
+		wget --no-hsts --progress=dot --tries=5 -P ${TARBALLS_DIR} -c $1 || wget --no-hsts --progress=dot --tries=5 --passive-ftp -P ${TARBALLS_DIR} -c $1 || \
 		(cd $TARBALLS_DIR; sh $TOP_DIR/glibc-get.sh $1; )
 		;; 
 	*)
 		# Note: if you need to use a proxy, try
 		# export http_proxy=<proxy_host>:<port>
-		wget --tries=5 -P ${TARBALLS_DIR} -c $1 || wget --tries=5 --passive-ftp -P ${TARBALLS_DIR} -c $1
+		wget --no-hsts --progress=dot --tries=5 -P ${TARBALLS_DIR} -c $1 || wget --no-hsts --progress=dot --tries=5 --passive-ftp -P ${TARBALLS_DIR} -c $1
 		# FIXME: support curl on systems that don't have wget
 		;;
 	esac
